### ISSUE TITLE
Show correct order for save masters with ESLs

### ISSUE THREAD

**Issue by nephatrine** [wrye-bash/wrye-bash#426]
In save files for FO4 (and presumably SSE), ESL files seem to be listed last (perhaps because they are FExxxxxx in-game) even though that doesn't correspond to the actual load order. Unfortunately, bash seems to just blindly compare the list from the save with the current LO and so marks all the save games as having the wrong load order when using ESL files and other mods.

**Comment by Utumno** [comment]
Well this is a known issue - unfortunately, the saves list all esls together plus there is this mystery surrounding esls - at the very least we need new code to compare the esls save order with the load order the game uses or imposes and is mirrored on teh LO displayed by bash - this all needs a specification and code

**Comment by Sharlikran** [comment]
@nephatrine I have discussed this with the Script Extender team they were not uninterested but didn't fully understand that you cannot reconstruct an ordered list of plugins from one ordered list of plugins and one unordered list of plugins.

![image](https://user-images.githubusercontent.com/1912426/50173161-84ab6900-02ab-11e9-9cdc-eb6ddfbc3ad4.png)

That was even what Elminster said and they still didn't seem interested in trying to help with a solution.

**Comment by Utumno** [comment]
Where is that discussion @Sharlikran ?

**Comment by Sharlikran** [comment]
SKSE discord, why do you ask?

**Comment by Utumno** [comment]
To have a look at latest developments - I lost track, hopefully there is a more clear understanding of esls. Can you post me the link?

**Comment by Sharlikran** [comment]
There are no links. I don't use forums anymore you would have to search and wads through countless comments on Discord (which I'm not asking you to do) if you need to see it for yourself.

This is not really related to this topic since it's about save games. Sorry for being a bit off topic but you appear to be trying to ask about ESL flagged files themselves. I have linked that to you already. As it stands there is no longer any mystery to it. What I added to the xEdit docs about [ESL flagged files](https://tes5edit.github.io/docs/9-appendix.html#s_9-2) is all anyone needs to know, period.

**Comment by Utumno** [comment]
Thanks for the links @Arthmoor ,
 @Sharlikran 

My idea is that if I can have a _very clear_ idea of esl load order (as in how the game orders those and there are some questions remaining) then we could rearrange the esls we display in the save masters so they match the load order they should have - this should eliminate most warnings. So if a save depends on ccc esls (say) those have a standard lo IIUC. Moving those to proper positions they will most probably fall into their actual place in the LO.

In parallel, we should verify that our currently displayed LO in the mod tab matches the current understanding of esl loading by the game.

**Comment by Utumno** [comment]
Ok will open an issue based on these and feedback by @Arthmoor here: https://afkmods.iguanadons.net/index.php?/topic/5079-plugin-files-and-you-esmeslesp/&do=findComment&comment=174631

**Comment by Utumno** [comment]
Quoting from that post:

> As far as the list of files in the save game, there are apparently 2 lists. One for all the r﻿﻿egul﻿ar ﻿plug﻿ins ﻿and ﻿ano﻿ther﻿ for﻿ all﻿ of﻿ the﻿ ESL﻿ fla﻿gged﻿ fi﻿les.﻿﻿ I don't know the details of how those are handled internally. Someone who knows how to reverse engineer the game data would have to comment on that. The game somehow knows how to keep it straight though because load order has been working just fine with all the ESL flagging everyone has been doing.

I would like some test:

- does an esl flagged file (either .esm or .esp) end up in that block?
- do esls in that block keep their __relative__ order? So if one makes a save with two esls (not hardcoded) and then flips those are those flipped in the new save?

What I would suggest is a new menu item "reorder esls" that would try and put the esls back to their positions. If a user only has ccc esls, for instance, they would end up in their hardcoded positions and this would make the save purple in 99% of the cases. All with a nice dialog as in "enabling that feature is experimental cause we can;t reconstruct the list from the limited data we have see pic by ElminsterEU :P"

------

> Bashed patches can also process ESL flagged files just fine, but that should not be allowed to include any CELL records that originate with an ESL flagged files in SSE or the cells will not load correctly in the game. So far this issue doesn't seem to exist with FO4.

Now that's quite another beast. The BP code I guess naturally expects 255 slots for plugins - if those suddenly raise to 600 I really don't know if the code is ready to cope. Any info on this appreciated and probably we need a new issue

**Comment by Utumno** [comment]
@Sharlikran @Arthmoor can any of you do the tests above for save files?

**Comment by Utumno** [comment]
> No need to test. Yes, any ESL flagged file is mapped into the FE slot by the engine. Doesn't matter what other flags it has on it. So they will be on that second load list in the save.

I could imagine the game just looking at the esl _extension_ though - _are esm/esp esls in that block?_ Especially esms. That would surprise me in a nice way and I would expect a couple bugs in bash's handling of those (it was a bit foggy back then when I wrote the save handling for esls)

> Yes, ESL flagged files keep their relative load order. 

again I am worried more about esl flagged epms

> You should keep in mind though that if the plugins.txt file is entirely lost, it will not be possible to reassemble the load order based only on what the save can tell you

Yes of course - what this will offer is a reordering of hardcoded esls (or rather ESL flagged if indeed espms end up there) and probably an attempt to put the rest in the places they correspond in the current load order- with a big fat warning that this is probably not true when the save was made. Or we could automatically order ccc esls to get rid of the most usual case and have a menu "Reorder esls" that does this for the rest. Needs code warning

**Comment by Utumno** [comment]
> I would suggest not trying to screw around with the official CC files in any way. The engine is going to load them its own way and will ignore what you tell plugins.txt to do with them. 

Certainly - what I meant is moving them in correct load order in the save masters list displayed

**Comment by Infernio** [comment]
This will be fixed once my cosaves rework is done and merged. As of https://github.com/wrye-bash/wrye-bash/commit/6f16d428701eabf6f4d3e237e7f08fa4e6090f49, it almost works correctly - the saves still show up as orange, but the master list is correct. Note that SKSE 2.0.15+ is required for this to work.

**Comment by Utumno** [comment]
That's marvelous - as I pointed out (but have not looked into it) a workaround that should suffice for most cases (and FO4) would be to call `set_load_order` with dry run which effectively corrects load order - this will at least move the esls to the master block (it does not matter where they are in the save file the game will load them that way). So we display the load order as it is and for most users this will make the save load order as similar as needed (but not more) to the actual/current load order

**Comment by Infernio** [comment]
The best course of action here would probably be to show a label in the GUI saying `u'The order shown below may not be correct. Install the most recent version of %s to fix this in future saves.' % bush.game.se.se_abbrev` if there are ESLs in the save master and we don't have a suitable cosave.

Would result in something like this:
 - The order shown below may not be correct. Install the most recent version of SKSE64 to fix this in future saves.
 - The order shown below may not be correct. Install the most recent version of F4SE to fix this in future saves.

**Comment by Utumno** [comment]
As I said above we should also try using the `set_load_order` API to correct most cases

**Comment by Infernio** [comment]
That'll work for ordering `.esl` files before regular files, but ESL-flagged `.esp` files would still show after all other `.esp` files.

**Comment by Utumno** [comment]
Not sure have to check - I think we use is_esl or similar that should check the flag

**Comment by Infernio** [comment]
I can't find a good way to do this, `load_order.py` doesn't seem to expose the necessary methods / fields(?)
If I want to use `set_load_order`, I'd need access to `load_order._game_handle`, right?
Things I've tried (all of these edit the last line in `bosh.SaveInfo.get_masters`):

1.
```python
fallback_masters = self.header.masters[:]
new_masters = load_order._game_handle.set_load_order(
    lord=fallback_masters, active=None, dry_run=True,
    previous_active=fallback_masters)[0]
return new_masters
```
I guess this is what you meant? This will move the CC files in the correct place, move up all `.esl` and ESLM-flagged `.esp` files to the end of the master block, and the other ESLs will remain at the bottom. It seems to also append any newly added mods to the master list though.

2.
```python
return load_order.get_ordered(self.header.masters)
```
This will adjust the masters to match the current load order as much as possible. Obviously not correct, since this will also reorder the masters to match any load order adjustments I've made since then.

**Comment by Utumno** [comment]
it's approach number one that I would have in mind, indeed not order the
mods just move esls and cc to their correct place - we would need to expose
a method in load_order.py like `correct_esl_load_order` that wraps the
set_load_order or possibly some of the Game.fix*** methods

**Comment by Sharlikran** [comment]
This can be closed. There is no way as mentioned to use the save game header to patch together two independent lists as shown in my screen shot. Elminster's summary should be sufficient.

Anyone with the latest F4SE or SKSE64 installed can already take advantage of the kind gesture that the script extender team offered us. The header now contains the order of plugins at runtime including ESL files. Wrye Bash already uses this to display the proper order.

None of the old file made with older versions of the script extender will work since they do not have the new plugin block.

**Comment by Infernio** [comment]
Sharlikran is correct, the main work for this is done as of 822c0bd16e5788d1811449810f34edfee16d77a1 and trying to display anything more is difficult engineering-wise and misleading. The only reason this issue is still open is because I want to add a warning to the Saves tab since this question keeps coming up.

**Comment by Utumno** [comment]
Hmm I still believe that smoothing out the display correcting the obvious serro\rs would work most of the time and it shouldn't be too hard - but yes it might prove tricky after all

**Comment by Infernio** [comment]
I think the warning gives a better user experience and points users to the right solution - installing SKSE/F4SE. It also seems to be much easier to implement :)

**Comment by Infernio** [comment]
![MastersWarning](https://user-images.githubusercontent.com/2845240/77953169-200ff400-72cd-11ea-99d3-ad6c728fad3a.png)

**Comment by Infernio** [comment]
This is what it looks like now (don't worry about all those missing mods, just some random savegame I grabbed from the Nexus):

![MastersWarning](https://user-images.githubusercontent.com/2845240/77953491-ac221b80-72cd-11ea-9979-91e6fcdb9a41.png)
